From 02da9b68334e444b1b49b2bac6e09fdb41208e34 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Wed, 12 Nov 2014 19:08:12 -0500
Subject: [infinband] cxgb4: Fix ntuple calculation for ipv6 and remove duplicate line

Message-id: <1415819300-32670-12-git-send-email-svemuri@redhat.com>
Patchwork-id: 99875
O-Subject: [RHEL7.1 PATCH BZ 1163467 11/19] RDMA/cxgb4: Fix ntuple calculation for ipv6 and remove duplicate line
Bugzilla: 1163467
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

This fixes ntuple calculation for IPv6 active open request for T5
adapter.  And also removes an duplicate line which got added in commit
92e7ae71726c ("iw_cxgb4: Choose appropriate hw mtu index and ISS for
iWARP connections")

Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit da22b896b119001b0cd9eb9ecb5b3fd85e948e43)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index 301b84d..353e75d 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -674,6 +674,7 @@ static int send_connect(struct c4iw_ep *ep)
  if (is_t5(ep->com.dev->rdev.lldi.adapter_type)) {
   opt2 |= T5_OPT_2_VALID;
   opt2 |= V_CONG_CNTRL(CONG_ALG_TAHOE);
+  opt2 |= CONG_CNTRL_VALID; /* OPT_2_ISS for T5 */
  }
  t4_set_arp_err_handler(skb, ep, act_open_req_arp_failure);
 
@@ -719,8 +720,6 @@ static int send_connect(struct c4iw_ep *ep)
  } else {
   u32 isn = (prandom_u32() & ~7UL) - 1;
 
-  opt2 |= T5_OPT_2_VALID;
-  opt2 |= CONG_CNTRL_VALID; /* OPT_2_ISS for T5 */
   if (peer2peer)
    isn += 4;
 
@@ -762,10 +761,10 @@ static int send_connect(struct c4iw_ep *ep)
    t5_req6->peer_ip_lo = *((__be64 *)
       (ra6->sin6_addr.s6_addr + 8));
    t5_req6->opt0 = cpu_to_be64(opt0);
-   t5_req6->params = (__force __be64)cpu_to_be32(
+   t5_req6->params = cpu_to_be64(V_FILTER_TUPLE(
        cxgb4_select_ntuple(
       ep->com.dev->rdev.lldi.ports[0],
-      ep->l2t));
+      ep->l2t)));
    t5_req6->rsvd = cpu_to_be32(isn);
    PDBG("%s snd_isn %u\n", __func__,
         be32_to_cpu(t5_req6->rsvd));
-- 
1.7.1